### PR TITLE
Any filesystem (SD/SPIFFS/LittleFS/SD_MMC)

### PR DESCRIPTION
**Pull request by tobozo**
---------------------------------------

### Discuss this PR [here](https://github.com/tobozo/esp32FOTA/discussions/5)

---------------------------------------

Followup for #25, #47, #60, #74, #79, #91

**Because this PR creates many new logical situations and I have only tested a few, this should be considered as untested code.**

Other changes mentioned in #25 are still being backported and tested.

### Functional changes

With this PR:
- SPIFFS is removed from the library core and replaced with an optional `fs::FS` filesystem abstraction.
- JSON Manifest accepts new keys `spiffs`, `littlefs`, `fatfs` holding path to filesystem image
- When one of the new keys is set in the manifest, execOTA will also flash the spiffs partition with the provided filesystem image (which can also be signed)

**:warning:  ESP32FOTA should detect when `#include <LittleFS>` (on any other filesystem) is included from the sketch, and automatically use that filesystem to verify certificate signatures. If no filesystem is detected, firmware signature check will always fail!**

```C++
// include filesystem first !

//#include <SD.h>
//#include <SD_MMC.h>
//#include <SPIFFS.h>
#include <LittleFS.h>
//#include <PSRamFS.h>
#include <esp32fota.h>

```

Sketches using more than one filesystem or compiled with older versions of platformio or may need this new method: 

```C++
    esp32FOTA.setCertFileSystem( &LittleFS ); // works with &SD, &SD_MMC, &SPIFFS
```

Also a new set of helper objects were added to handle the crypto assets in mixed situations:

```C++
// pick pubkey/rootca from filesystem
CryptoFileAsset( const char* path, fs::FS* filesystem ); 
// pick pubkey/rootca from progmem
CryptoMemAsset( const char* name, const char* bytes, size_t len );
```

Crypto asset usage with progmem: 

```C++

// could also be a byte array
const char* rsa_key_pub = R"RSA_KEY(
-----BEGIN PUBLIC KEY-----
BlahBlahBlah ... My RSA Key ... BlahBlahBlah
BlahBlahBlah ... My RSA Key ... BlahBlahBlah
BlahBlahBlah ... My RSA Key ... BlahBlahBlah
-----END PUBLIC KEY-----
)RSA_KEY";

CryptoMemAsset *MyRSAKey = new CryptoMemAsset("RSA Public Key", rsa_key_pub, strlen(rsa_key_pub)+1 );

void setup()
{
     // ...
    esp32FOTA.setPubKey( MyRSAKey );

}

```

Crypto asset usage with filesystem (can be different for each file): 

```C++
CryptoAsset* MyRSAKey = new CryptoFileAsset( "/path/to/public.key", &LittleFS  );
CryptoAsset* MyRootCA = new CryptoFileAsset( "/path/to/root_ca", &SD  );

void setup()
{
    // ....
    esp32FOTA.setPubKey( MyRSAKey );
    esp32FOTA.setRootCA( MyRootCA );
}
```

### Other changes in the PR

- This directive was added to prevent the latest espressif package from complaining at compilation:
```C++
  #pragma GCC diagnostic ignored "-Wmissing-field-initializers"
```
- new private method `debugPayloadVersion( const char* label, semver_t* version )` was added to remove redundancy in the code (also adds clarity and saves a few bytes)
- @tuan-karma's compiler fix for semver was applied
- cleaned up the includes list (compiles faster)
- added example based on LittleFS

@chrisjoyce911 may I suggest you enable the Discussions tab of this repository and create a few discussion threads such as:
- how to properly test the library (e.g. unit tests, github workflows, live/preprod tests, fixtures)
- complex update scenarios (e.g. gz/tar/tgz, firmware.gz, firmware+spiffs.tar, json+firmware+spiffs.tgz)

### CODE REVIEW

File: examples/anyFS/test/stage1/firmware.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
+[{
+    "type": "esp32-fota-http",
+    "version": 1,
+    "host": "github.com",
+    "port": 443,
+    "bin": "/tobozo/esp32FOTA/blob/tests/examples/anyFS/test/stage1/1.nosecurity.ino.bin?raw=true"
```
Comment (chrisjoyce911) [inline examples/anyFS/test/stage1/firmware.json:6]: Could we reference the main repo please

File: examples/anyFS/anyFS.ino [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  Serial.begin(115200);
+  // Provide filesystem with root_ca.pem to validate server certificate
+  if( ! LittleFS.begin( false ) ) {
+    Serial.println("LittleFS Mounting failed, aborting!");
+    while(1) vTaskDelay(1);
+  }
+  // use this when more than one filesystem is used in the sketch
+  // esp32FOTA.setCertFileSystem( &SD );
+
+  esp32FOTA.checkURL = "https://github.com/tobozo/esp32FOTA/raw/tests/examples/anyFS/test/stage1/firmware.json";
```
Comment (chrisjoyce911) [inline examples/anyFS/anyFS.ino:76]: Could we reference the main repo please

**Review by chrisjoyce911 — CHANGES_REQUESTED** [review]
Just a few small changes

File: examples/anyFS/test/stage1/firmware.json (hunk 1 above)
Comment (tobozo) [reply examples/anyFS/test/stage1/firmware.json:6]: - Solution 1: Deleting the `examples/anyFS/test/stage1` folder and updating the URLs in the ino files to `http://server/fota/fota.json` lorem ipsum values
- Solution 2: Moving the `examples/anyFS/test/stage1` folder outside the main tree (e.g. to `wiki` or `gh-pages` branch) and have the firmware.json and binaries generated by a workflow.

Comment (tobozo) [reply examples/anyFS/test/stage1/firmware.json]: I went with solution 1), the solution 2) can be implemented later

Comment (chrisjoyce911) [reply examples/anyFS/test/stage1/firmware.json]: Another option is setting up a custom domain for it .

I have a local system that lets me manage updates via a WebServer and DB , maybe I expand that to allow for a simple login and let people use it as  download source , and example bins could be stored there  ?

Comment (tobozo) [reply examples/anyFS/test/stage1/firmware.json]: Locally hosted system sounds like too much honey for the simple needs of a test suite. 

With the scope of this library covering C3, S2 and S3 models of ESP32, the different partition schemes, edge cases and experiments, the amount of logical situations to handle would explode to an unmanageable level.

The test suite is composed of 8 binary files and 1 json manifest, so exhaustively handling the whole test suite for *just one partition scheme* would require to create/compile 32 files.

I'm convinced something can be done with `workflow_dispatch` to mitigate this number.

- Use mandatory workflow dispatch form entries `esp32_model` `test_suite_url` and `test_suite_base path`
- Use `$test_suite_url` and `$test_suite_base path` to generate a custom firmware.json
- For each test-suite sketch (github matrix run):
  - Use `$test_suite_url` and `$test_suite_base path` to generate config.h
  - Compile sketch using `$esp32_model` as architecture (and optionnally `$partition_scheme`?)
  - Save the compiled binaries (may also have a spiffs binary) as artefacts
- Gather all artefacts from the actions tab
- Deploy manually on personal/local webserver

The last step of the workflow can eventually be replaced by a more specific action, but deploying is at the limit between test and production contexts and it feels like this workflow shouldn't oversetp.

**Comment by tobozo** [comment]
sorry about the spam, for some reason I can't test a dispatched workflow on my own fork and this is mildly infuriating

**Comment by tobozo** [comment]
Here's the deal for that new dispatched workflow:

image

:warning: Workflow must be triggered manually after editing the default values for each field in the highlighted part of capture, other fields can work fine with their respective default values.

What this workflow does:

- Build binaries for all examples in `anyFS/test` folder
- Fetch Root CA and populate spiffs/littlefs/progmem if applicable
- Create a JSON Manifest linking to every binary in the test suite
- Bundle the compiled binaries and the manifest into a downloadable zip archive

The resulting zip archive is only available from the github action tab though, so publishing the binaries to the web server should be done manually.

**Comment by tobozo** [comment]
@chrisjoyce911 is there anything else that need to be resolved before this is squash & merged ?

**Review by tobozo — COMMENTED** [review]
resoved
